File: Data Structures & Algorithms/search-insert-position/submission-3.kt
class Solution {
    fun searchInsert(nums: IntArray, target: Int): Int {

     fun binarySearch(left: Int, right: Int): Int {
        if(left > right) return left

        val mid = left + (right - left) / 2

        return if(target == nums[mid])
            mid
        else if(target > nums[mid])
            binarySearch(mid + 1, right)
        else
            binarySearch(left, mid - 1)
     }  

     return binarySearch(0, nums.lastIndex) 
    }
}

File: Data Structures & Algorithms/sqrtx/submission-3.cpp
class Solution {
   public:
    int mySqrt(int x) {
        int left = 0;
        int right = x;
        int result = 0;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            long long y = (long long) mid * mid;
            if (y > x) {
                right = mid - 1;
            } else if (y < x) {
                left = mid + 1;
                result = mid;
            } else {
                return mid;
            }
        }

        return result;
    }
};